src/spikeinterface/core/waveform_tools.py

```diff
@@ -397,6 +398,280 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
+            nbefore,
+            nafter,
+            return_scaled,
+            mode,
+            sparsity_mask,
+        )
+        if job_name is None:
+            job_name = f"extract waveforms {mode} mono buffer"
+
+        processor = ChunkRecordingExecutor(recording, func, init_func, init_args, job_name=job_name, **job_kwargs)
+        processor.run()
+
+    if mode == "memmap":
+        return all_waveforms
+    elif mode == "shared_memory":
+        if copy:
+            if shm is not None:
+                # release all sharedmem buffer
+                # empty array have None
+                shm.unlink()
+            return all_waveforms.copy()
+        else:
+            return all_waveforms, wf_array_info
+
+
+def _init_worker_distribute_single_buffer(
+    recording, unit_ids, spikes, wf_array_info, nbefore, nafter, return_scaled, mode, sparsity_mask
+):
+    worker_ctx = {}
+    worker_ctx["recording"] = recording
+    worker_ctx["wf_array_info"] = wf_array_info
+    worker_ctx["spikes"] = spikes
+    worker_ctx["nbefore"] = nbefore
+    worker_ctx["nafter"] = nafter
+    worker_ctx["return_scaled"] = return_scaled
+    worker_ctx["sparsity_mask"] = sparsity_mask
+    worker_ctx["mode"] = mode
+
+    if mode == "memmap":
+        filename = wf_array_info
+        all_waveforms = np.load(str(filename), mmap_mode="r+")
+        worker_ctx["all_waveforms"] = all_waveforms
+    elif mode == "shared_memory":
+        from multiprocessing.shared_memory import SharedMemory
+
+        shm, shm_name, dtype, shape = wf_array_info
+        shm = SharedMemory(shm_name)
+        all_waveforms = np.ndarray(shape=shape, dtype=dtype, buffer=shm.buf)
+        worker_ctx["shm"] = shm
+        worker_ctx["all_waveforms"] = all_waveforms
+
+    # prepare segment slices
+    segment_slices = []
+    for segment_index in range(recording.get_num_segments()):
+        s0 = np.searchsorted(spikes["segment_index"], segment_index)
+        s1 = np.searchsorted(spikes["segment_index"], segment_index + 1)
+        segment_slices.append((s0, s1))
+    worker_ctx["segment_slices"] = segment_slices
+
+    return worker_ctx
+
+
+# used by ChunkRecordingExecutor
+def _worker_distribute_single_buffer(segment_index, start_frame, end_frame, worker_ctx):
+    # recover variables of the worker
+    recording = worker_ctx["recording"]
+    segment_slices = worker_ctx["segment_slices"]
+    spikes = worker_ctx["spikes"]
+    nbefore = worker_ctx["nbefore"]
+    nafter = worker_ctx["nafter"]
+    return_scaled = worker_ctx["return_scaled"]
+    sparsity_mask = worker_ctx["sparsity_mask"]
+    all_waveforms = worker_ctx["all_waveforms"]
+
+    seg_size = recording.get_num_samples(segment_index=segment_index)
+
+    s0, s1 = segment_slices[segment_index]
+    in_seg_spikes = spikes[s0:s1]
+
+    # take only spikes in range [start_frame, end_frame]
+    # this is a slice so no copy!!
+    i0 = np.searchsorted(in_seg_spikes["sample_index"], start_frame)
```

Comment:
```suggestion
    i0 = np.searchsorted(in_seg_spikes["sample_index"], start_frame, side="right")
```
Shouldn't this be like this?

Reply:
I think it is side="left" so if the spike is exactly start_frame I want it in this buffer
(and if spike is exactly end_frame it is for next buffer)

let me double check. But then the error would be here since very very long time....

Reply:
I think you are correct, Sam. If you want the start_frame of the chunk in the buffer it should be left.
On the other hand, to following slicing convention maybe end_frame should be -1 in i1 just below?